### DISCUSSION TITLE
Unable to verify Minecraft jars

### DISCUSSION THREAD

**Discussion by arek-kwapis** [Abstrat-Technologies/rustedanvil#17] [Technical]
I need somewhere to bounce ideas. I've been working on trying to find more broken code and I successfully have - the Minecraft jar download function. However, I cannot for the life of me figure out why it fails.

I've gone as far as to diff my working copy of fml.py but that function is one of the ones I never touched. I can't for the life of me figure out why it fails now, but hasn't before. It makes no sense to me?

**Comment by arek-kwapis** [comment]
Okay, this is getting creepy... The server jar, usually the problematic one, does not error. Client does. What has changed? Why?

**Comment by Unloaded123456** [comment]
Is the problem about checksum validation? You should be able to edit MD5 checksums in the .py / .cfg files, there are online tools for generating MD5 hashes. 

If there is a HTTP 403 response, it is usually solved by replacing "http://" with "https://"

**Comment by arek-kwapis** [comment]
@Unloaded123456 honestly, I have no idea. The error is generic (because of course it is) and the files provided match exactly the ones I used originally. The hashes match what forge expects too. 

Now, forge's hash calculation algorithm is completely and utterly fucked beyond repair, which is why I rewrote it, but here's the funny thing - it downloaded the server jar just fine. That was always the issue. The client jar has always matched what forge expects, it was the server I had to wrangle. This time, they both match, AND it downloaded the server just fine, and the server and client jars are stored next to each other.

So as to why its doing it I have no idea. Everything is perfect and as it should be, there's zero reason to fail.

I might just blitz through the code with a bazooka once I get back home and see just what is up, because it's really confusing me

**Comment by arek-kwapis** [comment]
Okay so "progress" has been made. And by progress i mean a step back: the server jar also fails now. For no reason other than that it can... 🤦‍♂️

**Comment by arek-kwapis** [comment]
I figured it out! It seems that buried in a call within a call within a call within a call was a... wait for it... call! towards the broken hashcheck algo. My new hashcheck algo fixed it. 

That. Was. Painful. To. Debug.

**Reply by Unloaded123456** [reply]
Debugging runtime errors are painful, for instance my mod had weird texture errors and I could not fix for a day!

**Reply by arek-kwapis** [reply]
> Debugging runtime errors are painful, for instance my mod had weird texture errors and I could not fix for a day!

They definitely are! I looked further into it, and my issue was that it was jumping the gun, and providing the wrong error for what went wrong... Forge _totally_ isn't utterly broken 😂
